File: Day-4: Arrays Part-IV/6) Longest Substring without repeat.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int left = 0, right = 0;
        Map<Character, Integer> map = new HashMap<>();
        int ans = 0;
        while(right < n)
        {
            char c = s.charAt(right);
            if(!map.containsKey(c) || map.get(c) < left)
            {
                map.put(c, right);
                ans = Math.max(ans, right - left + 1);
            }
            else
            {
                left = map.get(c) + 1;
                map.put(c, right);
            }
            right++;
        }
        return ans;
    }
}

File: Day-04: Arrays Part-IV/6) Longest Substring without repeat.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int left = 0;
        int right = 0;
        unordered_map<char, int> mp;
        int ans = 0;
        while(right < n)
        {
            if(mp.count(s[right]) == 0 || mp[s[right]] < left)
            {
                mp[s[right]] = right;
                ans = max(ans, right - left + 1);
            }
            else
            {
                left = mp[s[right]] + 1;
                mp[s[right]] = right;
            }
            right++;
        }
        return ans;
    }
};
